src/realm/object-store/sync/sync_session.cpp

```diff
@@ -326,24 +327,29 @@ SyncSession::handle_refresh(const std::shared_ptr<SyncSession>& session, bool re
 
 SyncSession::SyncSession(SyncClient& client, std::shared_ptr<DB> db, const RealmConfig& config,
                          SyncManager* sync_manager)
-    : m_config(config)
-    , m_db(std::move(db))
-    , m_flx_subscription_store([this](bool use_flx_sync) -> std::shared_ptr<sync::SubscriptionStore> {
-        if (!use_flx_sync) {
+    : m_config{config}
+    , m_db{std::move(db)}
+    , m_flx_subscription_store{}
+    , m_original_sync_config{m_config.sync_config}
+    , m_migration_store{[&](bool is_flexible_sync) -> std::shared_ptr<sync::MigrationStore> {
+        using MigrationState = sync::MigrationStore::MigrationState;
+        if (is_flexible_sync) {
+            // Make sure the migration info is cleared
+            auto migration_store = sync::MigrationStore::create(m_db, [](MigrationState) {});
+            migration_store->clear();
+
             return nullptr;
         }
 
-        return sync::SubscriptionStore::create(m_db, [this](int64_t new_version) {
-            util::CheckedLockGuard lk(m_state_mutex);
-            if (m_state != State::Active && m_state != State::WaitingForAccessToken) {
-                return;
-            }
-            // There may be no session yet (i.e., waiting to refresh the access token).
-            if (m_session) {
-                m_session->on_new_flx_sync_subscription(new_version);
+        return sync::MigrationStore::create(m_db, [this](MigrationState state) {
+            // Migration state changed - Update the configuration to
+            // match the new sync mode.
+            {
+                util::CheckedLockGuard lock(m_config_mutex);
+                update_sync_config(state == MigrationState::Migrated);
```

Comment:
Why remove it from the MigrationStore? I think it makes more sense there. You could update the config when MigrateToFlx is received (equivalent to when the callback is called with MigrationState::Migrated). As mentioned previously, I am not sure we really need the callback.

Reply:
Added the function back and the SyncSession has a copy of the migration store, even for FLX - we can change this in the future if it doesn't make sense to keep it for native FLX.